### PR TITLE
Type annotations for sqlalchemy.sql.sqltypes

### PR DESCRIPTION
**Pull request by maxamly**
### Description

An attempt to annotate lib/sqlalchemy/sql/sqltypes.py with type hints (issue https://github.com/sqlalchemy/sqlalchemy/issues/6810)

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

**Comment by CaselIT** [comment]
Hi,

Thanks for the effort! We will review soon

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision fce2b1d27aa7abd6a3d8dadbbf788126ab7c89c9 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by CaselIT** [comment]
Hi, sorry for the delay.

Can you take a look at the errors?

```
pep484 installed: greenlet==2.0.2,mypy==1.1.1,mypy-extensions==1.0.0,SQLAlchemy @ file:///home/jenkins/workspace/pep484/sqlalchemy/sqlalchemy/.tox/.tmp/package/1/SQLAlchemy-2.0.7.dev0.zip,typing_extensions==4.5.0
pep484 run-test-pre: PYTHONHASHSEED='3227153439'
pep484 run-test: commands[0] | mypy ./lib/sqlalchemy
lib/sqlalchemy/sql/sqltypes.py:2760: error: No overload variant of "Numeric" matches argument types "int", "int", "bool"  [call-overload]
lib/sqlalchemy/sql/sqltypes.py:2760: note: Possible overload variants:
lib/sqlalchemy/sql/sqltypes.py:2760: note:     def [_N] __init__(self, precision: Optional[int] = ..., scale: Optional[int] = ..., decimal_return_scale: Optional[int] = ..., asdecimal: Literal[True] = ...) -> Numeric[Decimal]
lib/sqlalchemy/sql/sqltypes.py:2760: note:     def [_N] __init__(self, precision: Optional[int] = ..., scale: Optional[int] = ..., decimal_return_scale: Optional[int] = ..., asdecimal: Literal[False] = ...) -> Numeric[float]
Found 1 error in 1 file (checked 246 source files)
```

thanks!

**Comment by mlatysh** [comment]
Hello @CaselIT ! I've resolved the issues that were present due to some recent changes in the main branch. I have now rebased my branch and addressed the logical conflicts that emerged. Could you please initiate the tests and review the changes if everything is fine with tests?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 75db9eef943b48a11b9c01c7728296d4b5fb26ff of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 75db9eef943b48a11b9c01c7728296d4b5fb26ff added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4509

File: lib/sqlalchemy/sql/sqltypes.py [hunk 1]
```diff
@@ -155,15 +162,19 @@ class Indexable(TypeEngineMixin):
...
+            self, index: _ColumnExpressionArgument[_T]
+        ) -> ColumnElement[_T]:
             (
                 adjusted_op,
                 adjusted_right_expr,
                 result_type,
-            ) = self._setup_getitem(index)
+            ) = self._setup_getitem(
+                index
+            )  # type: ignore
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/sqltypes.py:177]: **mike bayer** (zzzeek) wrote:

fix return type for _setup_getitem, shouldnt need a type ignore here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4509

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

hey there -

I got through the first 200 lines and a large portion of the changes are incorrect.   It seems to be a long process for me to check and correct each change so I'm not sure how I want to proceed, if we can at least fix up the things I noted so far.   I think also it's important here that the patch adds *no* "type ignores" at all, because here we see they are only covering up incorrect typings in any case.    Better to leave the erroring typings in so that we can find and fix them more easily.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4509

**Comment by mlatysh** [comment]
Hello @zzzeek,

I have made the changes you requested and also addressed similar issues throughout the entire file. Additionally, I have implemented some further fixes and removed all instances of `type:ignore`s.

I would greatly appreciate it if you could review new changes and provide any additional instructions or feedback. Please let me know if there is anything else I can do to continue improving this PR.

Thank you!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 5860cc7248fd6106949b4f101edb76f8a8522132 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 5860cc7248fd6106949b4f101edb76f8a8522132 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4509

**Pull request closed without merging**